### PR TITLE
Fix for #70007, fallback to old implementation if optimized way fails

### PR DESCRIPTION
**Pull request by schuettecarsten**
Possible fix for #70007, fallback to old implementation if optimized way to query process name fails. This might happen if executed by a user that does not have the permission to call `QueryFullProcessImageNameW`.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70073) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70073) before we can accept your contribution.<br/><br/>:x: schuettecarsten [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70073)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=70073) it.</sub>

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (stephentoub) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: If this case will be common, the previous change was actually a deoptimization, as it'll now pay the cost of both approaches plus an exception throw/catch. I'm more inclined to see the previous change reverted, but I'll leave that up to @jeffhandley and @adamsitnik.

Comment (jaykrell) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: The new way usually works, is much more efficient (when it works),
and I believe also works sometimes when the old way does not, i.e. terminated processes.

I believe there is a race condition in the old way.
If you run a process that runs and exits quickly, you cannot
get its name. And .NET does not expose starting a process suspended (probably
because there is no way on Unix?) so there is no way to eliminate the race,
except the new way has no race.

Comment (jaykrell) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: Of course, .NET could always start suspended and get the name, costing memory and again, it is really very inefficient, the old way gets information including name for almost all processes on the system, only to compare by pid and pick out the one it wanted. The new way only gets the information for the one specified process.

Comment (jaykrell) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: For processes it start, .NET might also try to parse the name out of the command line.

**Comment by eerhardt** [comment]
Is there any way to write a test for this? If there is at least one Windows CI leg that runs as a non-admin user, we could have a test that finds the `services.exe` process ID (using native APIs or some other way so that `Process.ProcessName` isn't cached), and then calls

```C#
Process services = Process.GetProcessById(ServicesProcessId);
Assert.Equal("services", services.ProcessName);
```

**Comment by jaykrell** [comment]
The API to find services.exe etc. is the API used by the old code here -- it enumerates every process on the system and gives their process id, name, and more. So then you could turn around and feed that (just the process id) back into this code, have it fail the direct query, and get the exact same data again. Verifying that services.exe appears, or something?

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: This approach looks like an ideal compromise. From what I can see, the actual work is only done once as the result is cached in in `_processName`.

Comment (SteveDunn) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: Might it be worth narrowing down the exception that we expect?

Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: This approach looks like an ideal compromise. From what I can see, the actual work is only done once as the result is cached in in _processName.

Comment (jaykrell) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: Only once if you get the name yes.
I believe that is/was further mitigation around cost _and_ failure modes in the old way.
i.e. if you managed to get the name, in a race, prior to termination, then you could have it after termination.
The caching seems less well motivated now, in the usual path, but not much harm..

We really should get Windows fixed here.
There is little point imho in restricting access via `QueryFullProcessImageNameW` when the data is available with no extra permissions via NtQuerySystemInformation.

But I realize .NET has to support older Windows so fixing Windows, if it happens, does not immediately suffice.

Comment (schuettecarsten) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:888]: It's much easier - as `ProcessManager.GetProcessName` simply returns null in case it does not get a valid result from the kernel32 call, it is not neccessary to throw an exception and then catch it. I moved the fallback code to line 885.

**Comment by schuettecarsten** [comment]
> Is there any way to write a test for this?

Where to place such a test project?

**Comment by schuettecarsten** [comment]
I simplified the changes. The new code will fallback to the old implementation if `ProcessManager.GetProcessName` returns null. That's all, and that should be enough to fix the initial issue. So, no need to throw an exception and catch it again.

**Comment by ViktorHofer** [comment]
Can't directly answer that but what I know is that we have the `IsWindowsAndElevated` platform detection property that would be used to condition tests: https://github.com/dotnet/runtime/blob/3d74b00659fec817506e2888f87936518556e01c/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Windows.cs#L254.

**Comment by eerhardt** [comment]
@schuettecarsten - are you planning on adding a test for this?

**Comment by schuettecarsten** [comment]
@eerhardt - Is a test that relies on how it is executed in the build pipeline really a good idea? Of course we can analyze the process tree, look for the parents process name and check if it is services.exe. But how to figure out if it should be services.exe or can be something else?

So maybe the best solution is to go back to the bug that caused this fix and simply call `WindowsServiceHelpers.IsWindowsService` in a test, without checking the result itself. If the call works on all build pipelines, we are fine. Such a test already exists in `Microsoft.Extensions.Hosting`. And that this test never crashes on the build pipelines yet shows that there seems to be no pipeline that runs non-elevated.

**Comment by eerhardt** [comment]
> Of course we can analyze the process tree, look for the parents process name and check if it is services.exe. But how to figure out if it should be services.exe or can be something else?

This bug isn't about the parent process. It just so happens that the parent process was the way that Hosting was getting the `services.exe` process - that was the parent process when the app is executed as a Windows Service.

My thinking for a test would be:

* On Windows only, get the `services.exe` process ID using a mechanism that doesn't cause the process to be loaded into .NET's System.Diagnostics.Process cache/information. (One way to do this is p/invoke into Windows APIs. There may be others that I'm not aware of.)
* `Assert.Equal("services", Process.GetProcessById(servicesProcessId).ProcessName);`

> So maybe the best solution is to go back to the bug that caused this fix and simply call WindowsServiceHelpers.IsWindowsService in a test, without checking the result itself. If the call works on all build pipelines, we are fine. Such a test already exists in Microsoft.Extensions.Hosting. And that this test never crashes on the build pipelines yet shows that there seems to be no pipeline that runs non-elevated.

This is not an accurate assessment of the situation. Yes, there is a test in Microsoft.Extensions.Hosting that calls `IsWindowsService`. However, those tests are only ever executed outside of an actual Windows Service. The problem comes when `IsWindowsService` is called from a non-elevated process that IS running as a Windows Service. (Or more generally, a non-elevated process that was created from an elevated process.)

So maybe another way to test this is to spawn a non-elevated process from an elevated process, get the parent process, and check its name.

**Comment by schuettecarsten** [comment]
> So maybe another way to test this is to spawn a non-elevated process from an elevated process, get the parent process, and check its name.

I think this is the best solution. The issue has nothing to do with services, but with requesting details of the parent process from a non-evevated process. I will try to create such a test.

**Comment by schuettecarsten** [comment]
@eerhardt As a first step, I've created a simple console tool that reads the parent process name and writes it to the console. It uses the same code that IsWindowsService uses. I cannot make it compile as executable at the moment, the `OutputType` parameter is ignored or overwritten somewhere - any ideas how to fix this are welcome. Next, I have to create a test that starts this tool as an unelevated process and checks for the output.

**Comment by eerhardt** [comment]
> I've created a simple console tool that reads the parent process name and writes it to the console. It uses the same code that IsWindowsService uses. I cannot make it compile as executable at the moment, the OutputType parameter is ignored or overwritten somewhere - any ideas how to fix this are welcome. Next, I have to create a test that starts this tool as an unelevated process and checks for the output.

We don't need to make a new executable to do this. We have the `RemoteExecutor` which we use in tests to create a new process, if we need to execute some code in a separate process from the unit test process.

Here's an example:

https://github.com/dotnet/runtime/blob/475858d14c580a765c3e32cbb3f90cd5825e8b5c/src/libraries/System.Diagnostics.Process/tests/ProcessTests.Unix.cs#L240-L251

**Comment by schuettecarsten** [comment]
Unfortunately, I am not able to provide a test for this in a reasonable time (which means in July or August). So, please take a look is this fix is valid for a merge in the current state, maybe without the half test code, as I think it is too important to fix this before net7.0 is released.

**Comment by Jozkee** [comment]
@schuettecarsten did you try looking what's the error returned by QueryFullProcessImageName? Rather than doing a fallback, could it be possible that we are missing a permissions flag?

**Comment by jaykrell** [comment]
Other than not having to do with parent processes, this also has nothing to do with services. To write a test I suggest this, slightly icky but not terrible:
 Copy the fallback code.   
  Modify it to find the pid of services.exe.
 Assert there is at least one.
 Pass that pid to the “framework” and assert it works.
 Try it on “old” (7.0 prerelease) framework unelevated and see it fail.


Caveats:
 Services.exe is a windows implementation detail, and there could be others that already work here. It is just a name, not special or reserved or anything.
  Better would be to arrange to run a process, service or not, and verify a copy of the new code fails, fallback works, and framework works.

**Comment by Jozkee** [comment]
@jaykrell @eerhardt can you please take a look at this snippet of a possible test for this scenario? Since @schuettecarsten won't be able to provide it, I can send a commit to this PR with the proper changes, assuming it looks good to you.

```cs
[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsWindowsAndNotElevated))]
public void NonElevatedUser_QueryProcessNameOfSystemProcess()
{
    const string Services = "services";

    string currentProcessUser = Helpers.GetProcessUserName(Process.GetCurrentProcess());
    Assert.NotNull(currentProcessUser);

    Process? systemOwnedServices = null;

    foreach (var p in Process.GetProcessesByName(Services))
    {
        // returns the username of the owner of the process or null if the username can't be queried.
        // for services.exe, this will be null.
        string? servicesUser = Helpers.GetProcessUserName(p); 

        // this isn't really verifying that services.exe is owned by SYSTEM, but we are sure it is not owned by the current user.
        if (servicesUser != currentProcessUser)
        {
            systemOwnedServices = p;
            break;
        }
    }

    Assert.NotNull(systemOwnedServices);
    Assert.Equal(Services, systemOwnedServices.ProcessName);

    systemOwnedServices = Process.GetProcessById(systemOwnedServices.Id);
    Assert.Equal(Services, systemOwnedServices.ProcessName);
}
```

**Comment by Jozkee** [comment]
>  Rather than doing a fallback, could it be possible that we are missing a permissions flag?

FWIW, I looked at the native exception and OpenProcess returns Unauthorized Access, regardless of PROCESS_QUERY_LIMITED_INFORMATION. So, there's really no way to avoid regressing the scenario without falling back to NtQuerySystemInformation, AFAIK.

**Comment by eerhardt** [comment]
> can you please take a look at this snippet of a possible test for this scenario?

@Jozkee - can you try running that test without this fix? Does it fail with the same exception that we are getting in #70007?

```
Unhandled exception. System.InvalidOperationException: Process has exited, so the requested information is not available.
   at System.Diagnostics.Process.get_ProcessName()
   at ConsoleApp1.Program.Main(String[] args) 
```

**Comment by Jozkee** [comment]
@eerhardt yes, the first assert uses a `Process` obtained by `Process.GetProcessesByName("services.exe")`, which does have a cached `ProcessInfo`. the 2nd assert, uses `systemOwnedServices = Process.GetProcessById(systemOwnedServices.Id)`, which doesn't have the cache. Hence, the code without the fix does fail on the 2nd assert.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs
Comment (eerhardt) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:2575]: (nit) double blank lines

**Review by eerhardt — APPROVED** [review]
This LGTM - assuming the test you added failed previously with the same exception as the bug.

It would be nice if you could manually verify the Original Report in https://github.com/dotnet/runtime/issues/70007 using `WindowsServiceHelpers.IsWindowsService()`.

**Comment by Jozkee** [comment]
> It would be nice if you could manually verify the Original Report in https://github.com/dotnet/runtime/issues/70007 using WindowsServiceHelpers.IsWindowsService().

It is fixed, I created the service as stated in https://github.com/dotnet/runtime/issues/70007#issuecomment-1142606403, took the new dll from `.\artifacts\bin\testhost\net7.0-windows-Debug-x64\shared\Microsoft.NETCore.App\7.0.0` and place it on my `C:\Program Files\dotnet\shared\Microsoft.NETCore.App\7.0.0-preview.5.22301.12` folder.

**Comment by Jozkee** [comment]
CI issue is https://github.com/dotnet/runtime/issues/45868.

File: src/libraries/System.Diagnostics.Process/tests/Helpers.cs
Comment (jaykrell) [inline src/libraries/System.Diagnostics.Process/tests/Helpers.cs:55]: 8?

Comment (Jozkee) [reply src/libraries/System.Diagnostics.Process/tests/Helpers.cs:55]: that's [`TOKEN_QUERY`](https://docs.microsoft.com/en-us/windows/win32/secauthz/access-rights-for-access-token-objects). I didn't check what it exactly means since it was existing code, I just moved it here as I needed to reuse it.

**Pull request merged**
